src/Compilers/Core/Portable/FileSystem/PathUtilities.cs

```diff
@@ -773,6 +775,46 @@ public static bool IsValidFilePath([NotNullWhen(true)] string? fullPath)
         public static string NormalizeWithForwardSlash(string p)
             => DirectorySeparatorChar == '/' ? p : p.Replace(DirectorySeparatorChar, '/');
 
+        /// <summary>
+        /// Takes a normalized full path from <see cref="NormalizeWithForwardSlash(string)"/> and expands any '..' or '.'
+        /// into their equivalent representation.
+        /// </summary>
+        /// <returns>An equivalent path that does not contain any '..' or '.' path parts.</returns>
+        public static string ExpandNormalizedFullPathWithRelativeParts(string p)
+        {
+            if (!IsAbsolute(p))
+            {
+                // if this isn't an absolute path we can't expand it correctly without a root path
+                return p;
+            }
+
+            // GetPathParts also removes any instances of '.'
+            var parts = GetPathParts(p);
+
+            // On windows we need to skip the volume specifier, but remember it for re-joining later
```

Comment:
>volume specifier [](http://example.com/codeflow?start=46&length=16)

Are we testing volume specifiers? Consider having a unit test that calls `ExpandNormalizedFullPathWithRelativeParts()` directly.

Reply:
Yes, the test explicitly adds the volume for windows. Makes sense to add a direct test too though.